mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp

```diff
@@ -971,6 +981,42 @@ LogicalResult QuantizeLinearConverter::matchAndRewrite(
   return success();
 }
 
+LogicalResult
+ConvertConverter::matchAndRewrite(migraphx::ConvertOp op, OpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const {
+  if (!op.getZeroExtend()) {
+    rewriter.replaceOpWithNewOp<tosa::CastOp>(
+        op, getTypeConverter()->convertType(op.getResult().getType()),
+        adaptor.getInA());
+    return success();
+  }
+  // tosa.cast defaults to signed, which is what we usually want, but sometimes
+  // (currently this only happens when there's int4->int8 unpacking) we want the
+  // unsigned/zzero-extending cast.
+  Location loc = op.getLoc();
+  MIXRShapedType shapedResultType = op.getOutput().getType();
+  Type outElemType = shapedResultType.getElementType();
+  Type outType = getTypeConverter()->convertType(shapedResultType);
+  Value empty = rewriter.create<tensor::EmptyOp>(loc, outType,
+                                                 /*dynamicSize=*/ValueRange{});
+  SmallVector<AffineMap> iterationMaps(
+      2, rewriter.getMultiDimIdentityMap(shapedResultType.getRank()));
+  SmallVector<utils::IteratorType> iteratorKinds(shapedResultType.getRank(),
+                                                 utils::IteratorType::parallel);
+  auto converter = rewriter.create<linalg::GenericOp>(
```

Comment:
Please dont do this in MIGraphXToTosa. 
Please add a new conversion pass : MIGraphXToLinalg and do it there.

Alternatively why cant you use `tosa.rescale` ?

Reply:
(That is, strong disagree about not squeezing this in here)

Reply:
> I'm pretty sure those all are implicitly signed conversions
> 
> And ... I don't think tosa.rescale is the type change op, that's tosa.cast
> 
> ... Also ... like, we need to implement zero-extend _somehow_, and the only tosa way to do it is to add some unrealized casts to `ui4`.

https://github.com/llvm/llvm-project/blob/0b2d50c08d2eadf32f685fcd5b5468bc2ec7a16d/mlir/test/Conversion/TosaToLinalg/tosa-to-linalg.mlir#L1146

yep they do it by unrealized casts. They kind of have support for that but I agree its wonky.

> the Shaped -> Tensor type converter lives in MIGraphXToTosa

Clearly this does not need to live here. It could live somewhere common so we have more lowerings from MIGraphX dialect.

> that conversion needs to be done in one shot.

Not really.. its fine to partially convert it TOSA. you just need pull out the Shaped -> Tensor type converter

Reply:
Shaped -> Tensor type converter is using tensor dialect, I presume ?